### PR TITLE
Eliminate writes of non-dirty values in ScratchGlobalState

### PR DESCRIPTION
**Pull request by dwerner**

### CODE REVIEW

File: execution_engine/src/storage/global_state/scratch.rs [hunk 1]
```diff
@@ -30,26 +30,33 @@ use crate::{
 type SharedCache = Arc<RwLock<Cache>>;
 
 struct Cache {
-    stored_values: HashMap<Key, StoredValue>,
+    cached_values: HashMap<Key, (bool, StoredValue)>,
```
Comment (goral09) [inline execution_engine/src/storage/global_state/scratch.rs:33]: Alternatively, we could have separate cache for read-only and never write those back. I think something similar was done for `TrackingCopy`.

**Review by goral09 — APPROVED** [review]
So we're basically not writing back keys that were _read-only_ anymore?

**Review by Fraser999 — APPROVED** [review]

**Comment by dwerner** [comment]
> writing back keys that were _read-only_?

Yes, we were writing keys that had only been read.

**Comment by dwerner** [comment]
bors merge

**Comment by dwerner** [comment]
bors merge

**Pull request merged**